---
title: "WGCNA"
output: rmarkdown::github_document
---

# Load Libraries

```{r}
library(WGCNA)
library(DESeq2)
library(GEOquery)
library(tidyverse)
library(CorLevelPlot)
library(gridExtra)

allowWGCNAThreads() # for multi-threading
```

# 1. Fetch Data

```{r}
data <- read.delim('GSE152418_p20047_Study1_RawCounts.txt', header=T)

# get metadata
geo_id <- "GSE152418"
gse <- getGEO(geo_id, GSEMatrix=TRUE)
phenoData <- pData(phenoData(gse[[1]]))
head(phenoData)
```

```{r}
phenoData <- phenoData[,c(1,2,46:50)]

# prepare data
data[1:10, 1:10]
```

```{r}
data <- data %>%
  gather(key='samples', value='counts', -ENSEMBLID) %>%
  mutate(samples=gsub('\\.', '-', samples)) %>%
  inner_join(., phenoData, by=c('samples'='title')) %>%
  select(1,3,4) %>%
  spread(key='geo_accession', value='counts') %>%
  column_to_rownames(var='ENSEMBLID')
```

# 2. Outlier Detection

```{r}
# detect outlier genes
gsg <- goodSamplesGenes(t(data))
summary(gsg)
```

```{r}
gsg$allOK
```

```{r}
table(gsg$goodGenes)
```

```{r}
table(gsg$goodSamples)
```

```{r}
# remove genes that are detected as outliers
data <- data[gsg$goodGenes==TRUE,]
```

## 2.1 Hierarchical Clustering

```{r}
# detect outlier samples
htree <- hclust(dist(t(data)), method='average')
plot(htree)
```

## 2.2 PCA

```{r}
pca <- prcomp(t(data))
pca.dat <- pca$x

pca.var <- pca$sdev^2
pca.var.percent <- round(pca.var/sum(pca.var)*100, digits=2)

pca.dat <- as.data.frame(pca.dat)

ggplot(pca.dat, aes(PC1, PC2)) + 
  geom_point() + 
  geom_text(label=rownames(pca.dat)) + 
  labs(x=paste0('PC1: ', pca.var.percent[1], ' %'),
       y=paste0('PC2: ', pca.var.percent[2], ' %'))
```

*If your study design has a lot of replicates, make sure there aren't any batch effects present. If there are, you must correct for them before moving ahead.*

```{r}
# exclude outlier samples
samples.to.be.excluded <- c('GSM4615000', 'GSM4614993', 'GSM4614995')
data.subset <- data[,!(colnames(data) %in% samples.to.be.excluded)]
```

# 3. Normalization

```{r}
# exclude outlier samples
colData <- phenoData %>% 
  filter(!row.names(.) %in% samples.to.be.excluded)

# fix column names in colData
names(colData)
```

```{r}
names(colData) <- gsub(':ch1', '', names(colData))
names(colData) <- gsub('\\s', '_', names(colData))

# making the rownames and column names identical
all(rownames(colData) %in% colnames(data.subset))
all(rownames(colData) == colnames(data.subset))

# create dds
dds <- DESeqDataSetFromMatrix(countData = data.subset,
                              colData = colData,
                              design = ~ 1) # not specifying model
```

```{r}
## remove all genes with counts < 15 in more than 75% of samples (31*0.75=23.25)
## suggested by WGCNA on RNAseq FAQ

dds75 <- dds[rowSums(counts(dds) >= 15) >= 24,]
nrow(dds75)
```

```{r}
# perform variance stabilization
dds_norm <- vst(dds75)


# get normalized counts
norm.counts <- assay(dds_norm) %>% 
  t()
```

# 4. Network Construction

```{r}
# Choose a set of soft-thresholding powers
power <- c(c(1:10), seq(from = 12, to = 50, by = 2))

# Call the network topology analysis function
sft <- pickSoftThreshold(norm.counts,
                  powerVector = power,
                  networkType = "signed",
                  verbose = 5)

sft.data <- sft$fitIndices
```

```{r}
# visualization to pick power
a1 <- ggplot(sft.data, aes(Power, SFT.R.sq, label = Power)) +
  geom_point() +
  geom_text(nudge_y = 0.1) +
  geom_hline(yintercept = 0.8, color = 'red') +
  labs(x = 'Power', y = 'Scale free topology model fit, signed R^2') +
  theme_classic()


a2 <- ggplot(sft.data, aes(Power, mean.k., label = Power)) +
  geom_point() +
  geom_text(nudge_y = 0.1) +
  labs(x = 'Power', y = 'Mean Connectivity') +
  theme_classic()

grid.arrange(a1, a2, nrow = 2)
```

```{r}
# convert matrix to numeric
norm.counts[] <- sapply(norm.counts, as.numeric)

soft_power <- 18
temp_cor <- cor
cor <- WGCNA::cor
```

```{r}
# memory estimate w.r.t blocksize
bwnet <- blockwiseModules(norm.counts,
                 maxBlockSize = 14000,
                 TOMType = "signed",
                 power = soft_power,
                 mergeCutHeight = 0.25,
                 numericLabels = FALSE,
                 randomSeed = 1234,
                 verbose = 3)

cor <- temp_cor
```

# 5. Module Eigengenes

```{r}
module_eigengenes <- bwnet$MEs

# Print out a preview
head(module_eigengenes)
```

```{r}
# get number of genes for each module
table(bwnet$colors)
```

```{r}
# Plot the dendrogram and the module colors before and after merging underneath
plotDendroAndColors(bwnet$dendrograms[[1]], cbind(bwnet$unmergedColors, bwnet$colors),
                    c("unmerged", "merged"),
                    dendroLabels = FALSE,
                    addGuide = TRUE,
                    hang= 0.03,
                    guideHang = 0.05)
```

All genes that don't fall into other modules were assigned to the grey module.

# 6. Module Trait Association

```{r}
# create traits file - binarize categorical variables
traits <- colData %>% 
  mutate(disease_state_bin = ifelse(grepl('COVID', disease_state), 1, 0)) %>% 
  select(8)

# binarize categorical variables
colData$severity <- factor(colData$severity, levels = c("Healthy", "Convalescent", "ICU", "Moderate", "Severe"))

severity.out <- binarizeCategoricalColumns(colData$severity,
                           includePairwise = FALSE,
                           includeLevelVsAll = TRUE,
                           minCount = 1)

traits <- cbind(traits, severity.out)

# Define numbers of genes and samples
nSamples <- nrow(norm.counts)
nGenes <- ncol(norm.counts)

module.trait.corr <- cor(module_eigengenes, traits, use = 'p')
module.trait.corr.pvals <- corPvalueStudent(module.trait.corr, nSamples)

# visualize module-trait association as a heatmap
heatmap.data <- merge(module_eigengenes, traits, by = 'row.names')

head(heatmap.data)
```

```{r}
# convert 'Row.names' to row names (if necessary)
rownames(heatmap.data) <- heatmap.data$Row.names

# remove the 'Row.names' column, as it is now set as row names
heatmap.data <- heatmap.data[, -1]

# now heatmap.data should be ready for further analysis
head(heatmap.data)
```

```{r}
# view colNames in data
colnames(heatmap.data)
```

```{r}
# select module eigengene columns
eigengene_columns <- colnames(heatmap.data)[1:16]  # module eigengenes (1 to 16)

# select trait columns
trait_columns <- colnames(heatmap.data)[17:21]  # traits (17 to 21)

# generate the correlation level plot
CorLevelPlot(heatmap.data,
             x = trait_columns,  # traits
             y = eigengene_columns,  # module eigengenes
             col = c("blue1", "skyblue", "white", "pink", "red"))
```

```{r}
module.gene.mapping <- as.data.frame(bwnet$colors)
module.gene.mapping %>% 
  filter(`bwnet$colors` == 'turquoise') %>% 
  rownames()

module.gene.mapping
```

# 7. Intramodular Analysis

Identifying driver genes

```{r}
# calculate the module membership and the associated p-values

# the module membership/intramodular connectivity is calculated as the correlation of the eigengene and the gene expression profile. 
# this quantifies the similarity of all genes on the array to every module.

module.membership.measure <- cor(module_eigengenes, norm.counts, use = 'p')
module.membership.measure.pvals <- corPvalueStudent(module.membership.measure, nSamples)


module.membership.measure.pvals[1:10,1:10]
```

```{r}
# calculate the gene significance and associated p-values
gene.signf.corr <- cor(norm.counts, traits$data.Severe.vs.all, use = 'p')
gene.signf.corr.pvals <- corPvalueStudent(gene.signf.corr, nSamples)

gene.signf.corr.pvals %>% 
  as.data.frame() %>% 
  arrange(V1) %>% 
  head(25)
```

Using the gene significance you can identify genes that have a high significance for trait of interest.

Using the module membership measures you can identify genes with high module membership in interesting modules.
